buster/chatbot.py

```diff
@@ -217,36 +219,16 @@ def check_response_relevance(
         # Likely that the answer is meaningful, add the top sources
         return score < unk_threshold
 
-    def format_response(self, response: str, matched_documents: pd.DataFrame, text_after_response: str) -> str:
-        """
-        Format the response by adding the sources if necessary, and a disclaimer prompt.
-        """
-        sep = self.cfg.separator
-
-        is_relevant = self.check_response_relevance(
-            response=response,
-            engine=self.cfg.embedding_model,
-            unk_embedding=self.unk_embedding,
-            unk_threshold=self.cfg.unknown_threshold,
-        )
-        if is_relevant:
-            # Passes our relevance detection mechanism that the answer is meaningful, add the top sources
-            response = self.add_sources(
-                response=response,
-                matched_documents=matched_documents,
-                sep=self.cfg.separator,
-                format=self.cfg.link_format,
-            )
-
-        response += f"{sep}{sep}{sep}{text_after_response}{sep}"
-
-        return response
-
-    def process_input(self, question: str) -> str:
+    def process_input(self, question: str, formatter: Formatter = None) -> str:
         """
         Main function to process the input question and generate a formatted output.
         """
 
+        if formatter is None and self.cfg.link_format not in FORMATTERS:
+            raise ValueError(f"Unknown link format {self.cfg.link_format}")
+        elif formatter is None:
+            formatter = FORMATTERS[self.cfg.link_format]()
```

Comment:
would it make more sense to initialize the formatter in the `__init__`?

Reply:
Yes, but that's not what the current app does, and I'm trying to preserve
the interface to avoid merge conflicts with the other changes we are making